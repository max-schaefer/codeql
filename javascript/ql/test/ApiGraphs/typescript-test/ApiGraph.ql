/**
 * @name API graph
 * @description Shows the API graph of a code base.
 * @kind graph
 * @id js/api-graph
 */

import javascript

/** Gets a string representation of the location information for `nd`. */
string locationString(API::Node nd) {
  exists(string fp, int sl, int sc, int el, int ec | nd.hasLocationInfo(fp, sl, sc, el, ec) |
    result = fp + ":" + sl + ":" + sc + ":" + el + ":" + ec
  )
}

/**
 * Gets the rank of node `nd` among all nodes ordered by depth, string representation,
 * and location.
 */
int nodeRank(API::Node nd) {
  nd =
    rank[result + 1](API::Node nd2 |
      |
      nd2 order by nd2.getDepth(), nd2.toString(), locationString(nd2)
    )
}

/**
 * Gets the rank of `lbl` among all labels on outgoing edges of `pred`, ordered alphabetically.
 */
int labelRank(API::Node pred, string lbl, API::Node succ) {
  lbl + "->" + nodeRank(succ) =
    rank[result + 1](string l, API::Node s |
      s = pred.getASuccessor(l)
      or
      pred.refersTo(s) and l = ""
    |
      l + "->" + nodeRank(s)
    )
}

query predicate nodes(API::Node f, string key, string value) {
  key = "semmle.order" and
  value = nodeRank(f).toString()
}

query predicate edges(API::Node pred, API::Node succ, string key, string value) {
  exists(string lbl |
    succ = pred.getASuccessor(lbl)
    or
    pred.refersTo(succ) and
    lbl = ""
  |
    key = "semmle.label" and
    value = lbl
    or
    key = "semmle.order" and
    value = labelRank(pred, lbl, succ).toString()
  )
}
