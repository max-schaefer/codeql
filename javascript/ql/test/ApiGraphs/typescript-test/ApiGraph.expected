#-----| root
#-----| module mongoose -> use (module mongoose)
#-----| module typescript-test -> def (module typescript-test)
#-----| module typescript-test -> use (module typescript-test)

#-----| def (module typescript-test)
#-----| member exports -> def (member exports (module typescript-test))

#-----| use (module mongoose)
#-----| member exports -> use (member exports (module mongoose))

#-----| use (module typescript-test)
#-----| member exports -> use (member exports (module typescript-test))

#-----| def (member exports (module typescript-test))
#-----| member AccessLevel -> def (member AccessLevel (member exports (module typescript-test)))
#-----| member AccessLevel -> def (member AccessLevel (member exports (module typescript-test)))
#-----| member AccessLevel.Boss -> def (member AccessLevel.Boss (member exports (module typescript-test)))
#-----| member AccessLevel.None -> def (member AccessLevel.None (member exports (module typescript-test)))
#-----| member AccessLevel.Read -> def (member AccessLevel.Read (member exports (module typescript-test)))
#-----| member AccessLevel.ReadWrite -> def (member AccessLevel.ReadWrite (member exports (module typescript-test)))
#-----| member AccessLevel.Write -> def (member AccessLevel.Write (member exports (module typescript-test)))
#-----| member MyModel -> def (member MyModel (member exports (module typescript-test)))
#-----| member Sized -> def (member Sized (member exports (module typescript-test)))

#-----| use (member exports (module mongoose))
#-----| member Model -> use (member Model (member exports (module mongoose)))

#-----| use (member exports (module typescript-test))
#-----| member AccessLevel -> use (member AccessLevel (member exports (module typescript-test)))

index.ts:
#   10| def (member AccessLevel (member exports (module typescript-test)))

#-----| def (member AccessLevel (member exports (module typescript-test)))
#-----| member Boss -> def (member AccessLevel.Boss (member exports (module typescript-test)))
#-----| member None -> def (member AccessLevel.None (member exports (module typescript-test)))
#-----| member Read -> def (member AccessLevel.Read (member exports (module typescript-test)))
#-----| member ReadWrite -> def (member AccessLevel.ReadWrite (member exports (module typescript-test)))
#-----| member Write -> def (member AccessLevel.Write (member exports (module typescript-test)))

#-----| def (member AccessLevel.Boss (member exports (module typescript-test)))

#-----| def (member AccessLevel.None (member exports (module typescript-test)))

#-----| def (member AccessLevel.Read (member exports (module typescript-test)))

#-----| def (member AccessLevel.ReadWrite (member exports (module typescript-test)))

#-----| def (member AccessLevel.Write (member exports (module typescript-test)))

#-----| def (member MyModel (member exports (module typescript-test)))
#-----|  -> use (member Model (member exports (module mongoose)))

#-----| def (member Sized (member exports (module typescript-test)))

#-----| use (member AccessLevel (member exports (module typescript-test)))

#-----| use (member Model (member exports (module mongoose)))
